packages/core/mesh/edge-client/src/edge-ws-connection.ts

```diff
@@ -0,0 +1,148 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import WebSocket from 'isomorphic-ws';
+
+import { scheduleTask, scheduleTaskInterval } from '@dxos/async';
+import { Context, Resource } from '@dxos/context';
+import { invariant } from '@dxos/invariant';
+import { log, logInfo } from '@dxos/log';
+import { buf } from '@dxos/protocols/buf';
+import { MessageSchema, type Message } from '@dxos/protocols/buf/dxos/edge/messenger_pb';
+
+import { protocol } from './defs';
+import { type EdgeIdentity } from './edge-identity';
+import { toUint8Array } from './protocol';
+
+const SIGNAL_KEEPALIVE_INTERVAL = 5_000;
+
+export type EdgeWsConnectionCallbacks = {
+  onConnected: () => void;
+  onMessage: (message: Message) => void;
+  onRestartRequired: () => void;
+};
+
+export class EdgeWsConnection extends Resource {
+  private _inactivityTimeoutCtx: Context | undefined;
+  private _ws: WebSocket | undefined;
+
+  constructor(
+    private readonly _identity: EdgeIdentity,
+    private readonly _connectionInfo: { url: URL; protocolHeader?: string },
```

Comment:
I'd move the header formatting here as well and pass in the encoded credentail - seems like header format is closely releated to the websocket protocol